### ISSUE TITLE
Custom encoding/decoding similar to `encoding/json`

### ISSUE THREAD

**Issue by zegl** [rethinkdb/rethinkdb-go#363]
Hi.

With `encoding/json` it is possible to change how a field is encoded and decoded with the `Marshaler` and `Unmarshaler` interfaces.

I would like to have the same flexibility in `gorethink`, to make it easier to use custom datatypes.

An example of something that doesn't work today:

``` go
type CustomID [10]byte

type Item struct {
    ID CustomID `gorethink:"id"`
}
```

Currently `ID` would be saved as a list, which means that it can't be used as a primary key.

I have written an example ( https://play.golang.org/p/iZo9pJ-Yw1 ) with a suggestion of how this problem can be solved.

**Comment by dancannon** [comment]
GoRethink actually includes a (currently undocumented feature) which should do exactly what you would like! By implementing the [`Marshaler`](https://godoc.org/github.com/dancannon/gorethink/encoding#Marshaler) and [`Unmarshaler`](https://godoc.org/github.com/dancannon/gorethink/encoding#Unmarshaler) interfaces you can change how a type is encoded and decoded, however instead of converting to bytes the `Marshaler` should return `interface{}` which is encoded by the JSON encoder, this may be an issue for what you are trying to do but hopefully not.

There may also be another issue here as a byte array should be base64 encoded and stored as the `BINARY` ReQL type by GoRethink when inserting into RethinkDB. I will try to fix this issue and add documentation for the interfaces in the next release. Let me know if you have any questions about the `Marshaler` and `Unmarshaler` interfaces or if you think anything else should be changed.

**Comment by paralin** [comment]
@dancannon Really cool, this way I can actually translate my Protobuf GeoJSON esque data structures into RQL Geo structures. Cool stuff!

**Comment by zegl** [comment]
@dancannon Aha! The `Marshaler` and `Unmarshaler` interfaces solves the problem that I needed to get solved. 👍 

Thank you for your response and help.

**Comment by dancannon** [comment]
The [...]byte fix and improved documentation have been released as part of `v2.2.0`
